dips/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::str::FromStr;
+
+pub use alloy_core;
+pub use alloy_rlp;
+pub use alloy_signer;
+pub use alloy_sol_types;
+
+use alloy_core::primitives::Address;
+use alloy_rlp::{RlpDecodable, RlpEncodable};
+use alloy_signer::Signature;
+use alloy_sol_types::{sol, SolStruct};
+
+sol! {
+    // EIP712 encoded bytes, ABI - ethers
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct SignedIndexingAgreementVoucher {
+        IndexingAgreementVoucher voucher;
+        bytes signature;
+    }
+
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct IndexingAgreementVoucher {
+      // should coincide with signer
+        address payer;
+        // should coincide with indexer
+        address payee;
+         // data service that will initiate payment collection
+        address service;
+        // initial indexing amount max
+        uint256 maxInitialAmount;
+        uint256 maxOngoingAmountPerEpoch;
+        // time to accept the agreement, intended to be on the order
+        // of hours or mins
+        uint64 deadline;
+        uint32 maxEpochsPerCollection;
+        uint32 minEpochsPerCollection;
+        // after which the agreement is complete
+        uint32 durationEpochs;
+        bytes metadata;
+    }
+
+    // the vouchers are generic to each data service, in the case of subgraphs this is an ABI-encoded SubgraphIndexingVoucherMetadata
+    #[derive(Debug, RlpEncodable, RlpDecodable, PartialEq)]
+    struct SubgraphIndexingVoucherMetadata {
+        uint256 pricePerBlock; // wei GRT
+        bytes32 protocolNetwork; // eip199:1 format
+        // differentiate based on indexed chain
+        bytes32 chainId; // eip199:1 format
+    }
+}
```

Comment:
Note for the future: We might want this to be part of the `thegraph_core` crate to share the same code between the different parts.